lib/jekyll/collection.rb

```diff
@@ -210,7 +210,28 @@ def read_document(full_path)
       end
     end
 
-    private
+    def sort_docs
+      return docs.sort! unless metadata["sort_by"].is_a?(String)
+      sort_docs_by_key!
+    end
+
+    def sort_docs_by_key!
+      meta_key = metadata["sort_by"]
+      docs.sort_by! { |d| d.data[meta_key] }
+    rescue ArgumentError
```

Comment:
This is unacceptable to deal with `d.data[meta_key]`. A `nil` value must be dealt with in the `sort_by!` block.

Reply:
One way to handle this is to sort by an array of values:

```ruby
# sort by the key's value, then by its path if the first is the same.
docs.sort_by! { |d| [d.data[meta_key], d.path]  }
```

This would be my preferred approach.

```ruby
>> nil <=> nil
=> 0
>> nil <=> "hi"
=> nil
>> "hi" <=> nil
=> nil
```

Reply:
@parkr I think you missed the point why I rescued `ArgumentError` here:
An array of objects **cannot be sorted** by an attribute **if all constituent elements *do not* have the attribute defined**

So while your suggestion can be adapted to allow users to sort by multiple attributes, it'll fail if one of the attribute is not defined in all of the constituent documents..

Reply:
@ashmaroli Well, I just spent the last hour digging and digging and I found something interesting. Of course, we _do_ allow sorting by attributes which all constituent elements may not have in `Jekyll::Filters#sort`. We don't use `sort_by` there. We use `sort`. I was digging through the documentation earlier for `sort_by` and I came across a mention of a Schwartzian transform, which is used to speed up sort calls by caching values. I wrote a benchmark and updated `Jekyll::Filters#sort_input` to use this: https://github.com/jekyll/jekyll/pull/6342

Using `rescue ArgumentError` is using a `rescue` as a control flow mechanism. At the very least, we should be checking the properties using `return docs.sort! unless docs.all? { |d| d.data.key?(meta_key) }` instead of hoping that `sort_by` will always raise `ArgumentError`.

I'd propose we fall back to `path` if the key doesn't exist:

```ruby
meta_key = metadata["sort_by"]
docs.map! do |d| 
  Jekyll.logger.warn "Sort warning:", "Missing sort key '#{meta_key}' on document #{d.relative_path}"
  [d.data[meta_key], d]
end.sort! do |apple, orange|
  if !apple.first.nil? && !orange.first.nil?
    # If both documents have the property, sort by that property.
    puts "Sorting by the property"
    apple.first <=> orange.first
  elsif !apple.first.nil? && orange.first.nil?
    -1
  elsif apple.first.nil? && !orange.first.nil?
     1
  else
    # Fall back to Document#<=> if both documents don't have the property.
    apple.last <=> orange.last
  end
end.map! do |apple|
  apple.last # return the Document object
end
```